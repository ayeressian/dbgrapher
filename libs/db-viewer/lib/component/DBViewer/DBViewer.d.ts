type DbViewerEventMap = import("./event").DbViewerEventMap;
type Schema = import("../../schema").Schema;
type Viewport = import("../../store/view").Viewport;

export default class DbViewer extends HTMLElement {
  setSchema(schema: Schema, viewport: Viewport): void;
  getSchema(): Schema;

  addEventListener<K extends keyof DbViewerEventMap>(
    type: K,
    listener: (this: DbViewer, ev: DbViewerEventMap[K]) => unknown,
    options?: boolean | AddEventListenerOptions
  ): void;

  removeEventListener<K extends keyof DbViewerEventMap>(
    type: K,
    listener: (this: HTMLFormElement, ev: DbViewerEventMap[K]) => unknown,
    options?: boolean | EventListenerOptions
  ): void;
}

export * from "../../schema";
export * from "./event";
export { default as Point } from "../../point";
