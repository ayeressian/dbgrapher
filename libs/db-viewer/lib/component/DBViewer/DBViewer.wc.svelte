<svelte:options tag="db-viewer" />

<script lang="ts">
  import type { Schema } from "../../schema";
  import { onMount, setContext } from "svelte";
  import Table from "../Table/Table.svelte";
  import Relation from "../Relation/Relation.svelte";
  import style from "../../style.css";
  import type { Viewport } from "../../store/view";
  import applyStyle from "../../wc-style-applier";
  import type Point from "../../point";
  import Minimap from "../Minimap/Minimap.svelte";
  import { Store } from "../../store/store";
  import type { TableData } from "./event";
  import EventDispatcher from "./event-dispatcher";
  import type DbViewer from "./DBViewer";
  import Zoom from "../Zoom/Zoom.svelte";

  const store = new Store();
  setContext("store", store);

  const schemaStore = store.schema.schema;

  export const getSchema = (): Schema => store.schema.getSchema();
  export const setSchema = (schema: Schema, viewport: Viewport) => {
    store.schema.setSchema(schema, viewport);
  }

  const SCROLL_TO_ZOOM_MULTIPLIER = 0.01;
  const MAXIMUM_ZOOM = 0.4;
  const MINIMUM_ZOOM = 4;
  const DEFAULT_ZOOM = 1;

  let container: HTMLDivElement;
  let svgElem: SVGSVGElement;

  const zoomStore = store.view.zoom;
  const viewSizeStore = store.view.viewSize;
  const viewPosStore = store.view.viewPos;
  const viewSizeAfterZoomStore = store.view.viewSizeAfterZoom;
  const relationsStore = store.relation.relations;

  let eventDispatcher: EventDispatcher;

  onMount(() => {
    const shadowRoot = container.parentNode as ShadowRoot;
    applyStyle(shadowRoot, style);

    const host = shadowRoot.host as DbViewer;

    eventDispatcher = new EventDispatcher(host);

    store.view.viewSize.set({
      width: host.clientWidth,
      height: host.clientHeight,
    });

    const svgResizeObserver = new ResizeObserver((entries) => {
      store.view.viewSize.set({
        width: host.clientWidth,
        height: host.clientHeight,
      });
    });
    svgResizeObserver.observe(host);
    zoomStore.set(DEFAULT_ZOOM);
  });

  function getSvgMousePos(event: MouseEvent) {
    const pt = svgElem.createSVGPoint();
    pt.x = event.clientX;
    pt.y = event.clientY;
    const domPoint = pt.matrixTransform(svgElem?.getScreenCTM()?.inverse());
    return { x: domPoint.x, y: domPoint.y };
  }

  function zoomHandler(event: WheelEvent) {
    event.preventDefault();

    const bb = svgElem.getBoundingClientRect();
    const mousePos: Point = {
      x: event.clientX - bb.left,
      y: event.clientY - bb.top,
    };
    const zoomAmount = event.deltaY * SCROLL_TO_ZOOM_MULTIPLIER;
    const oldZoom = $zoomStore;
    let newZoom = $zoomStore + zoomAmount;
    newZoom = Math.max(newZoom, MAXIMUM_ZOOM);
    newZoom = Math.min(newZoom, MINIMUM_ZOOM);

    if (newZoom === oldZoom) {
      return;
    }

    const newWidth = $viewSizeStore.width * newZoom;
    const oldWidth = $viewSizeStore.width * oldZoom;

    const newHeight = $viewSizeStore.height * newZoom;
    const oldHeight = $viewSizeStore.height * oldZoom;

    const widthDelta = oldWidth - newWidth;
    const heightDelta = oldHeight - newHeight;

    viewPosStore.update((viewPosVal) => {
      viewPosVal.x =
        viewPosVal.x + (widthDelta * mousePos.x) / $viewSizeStore.width;
      viewPosVal.y =
        viewPosVal.y + (heightDelta * mousePos.y) / $viewSizeStore.height;
      return viewPosVal;
    });
    zoomStore.set(newZoom);
    eventDispatcher.onZoom({ newZoom, oldZoom, target: getSvgMousePos(event) });
  }

  function scrollOperation(scrollXAmount: number, scrollYAmount: number) {
    viewPosStore.update((viewPosVal) => {
      viewPosVal.x += scrollXAmount;
      viewPosVal.y += scrollYAmount;
      return viewPosVal;
    });
  }

  function scrollHandler(event: WheelEvent) {
    event.preventDefault();
    scrollOperation(event.deltaX, event.deltaY);
  }

  function mousewheel(event: WheelEvent) {
    if (event.ctrlKey) {
      zoomHandler(event);
      return;
    }
    scrollHandler(event);
  }

  function onTableMouseDown({ name }: TableData) {
    const index = $schemaStore.tables.findIndex((table) => table.name === name);
    const [table] = $schemaStore.tables.splice(index, 1);
    $schemaStore.tables = [...$schemaStore.tables, table];
  }

  function click(event: MouseEvent) {
    eventDispatcher.onViewportClick(getSvgMousePos(event));
  }
</script>

<div id="container" bind:this={container}>
  <svg
    class="view"
    bind:this={svgElem}
    on:wheel={mousewheel}
    on:click={click}
    version="2"
    xmlns="http://www.w3.org/2000/svg"
    viewBox={`${$viewPosStore.x} ${$viewPosStore.y} ${$viewSizeAfterZoomStore.width} ${$viewSizeAfterZoomStore.height}`}
  >
    {#each $relationsStore ?? [] as relation (relation.key)}
      <Relation
        {...relation}
        click={eventDispatcher.onRelationClick}
        dblClick={eventDispatcher.onRelationDblClick}
        contextMenu={eventDispatcher.onRelationContextMenu}
      />
    {/each}   
    {#each $schemaStore?.tables ?? [] as { name } (name)}
      <Table
        {name}
        mouseDown={onTableMouseDown}
        click={eventDispatcher.onTableClick}
        dblClick={eventDispatcher.onTableDblClick}
        contextMenu={eventDispatcher.onTableContextMenu}
      />
    {/each}     
  </svg>  
  <Minimap />
  <Zoom maxZoom={MAXIMUM_ZOOM} minZoom={MINIMUM_ZOOM}/>
</div>
