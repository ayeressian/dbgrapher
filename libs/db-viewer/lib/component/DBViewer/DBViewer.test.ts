import {
  cleanup,
  fireEvent,
  render,
  type RenderResult,
} from "@testing-library/svelte";
import DBViewer from "./DBViewer.wc.svelte";
import school from "../../../src/school";
import { beforeEach, describe, expect, it, vi } from "vitest";

describe(DBViewer.name, () => {
  let component: RenderResult;

  const viewWidth = 800,
    viewHeight = 1000;

  global.ResizeObserver = vi.fn().mockImplementation((func) => {
    func([{ contentRect: { width: viewWidth, height: viewHeight } }]);
    return {
      observe: vi.fn(),
    };
  });

  vi.mock("../../wc-style-applier", () => {
    return {
      default: vi.fn(),
    };
  });

  beforeEach(async () => {
    cleanup();
    component = render(DBViewer, {
      schema: school,
    });
  });
  it("should render properly", () => {
    expect(component).toMatchSnapshot();
  });

  describe("scroll view", () => {
    let svgElem: SVGElement;
    const deltaX = 100,
      deltaY = 200;
    beforeEach(() => {
      svgElem = document.querySelector("svg") as SVGElement;
      fireEvent.wheel(svgElem, { deltaX, deltaY });
    });
    it("viewport should change", () => {
      expect(svgElem.getAttribute("viewBox")).toEqual(
        `${deltaX} ${deltaY} ${viewWidth} ${viewHeight}`
      );
    });
  });

  describe("zoom view", () => {
    let svgElem: SVGElement;
    const clientX = 100,
      clientY = 200,
      deltaY = -20,
      cameraXResult = (clientX - (viewWidth * 0.3) / 2) * 0.2,
      cameraYResult = (clientY - (viewHeight * 0.3) / 2) * 0.2,
      viewWidthResult = viewWidth * 0.8,
      viewHeightResult = viewHeight * 0.8;

    beforeEach(() => {
      Object.defineProperty(
        global.SVGGraphicsElement.prototype,
        "getScreenCTM",
        {
          writable: true,
          value: vi.fn().mockReturnValue({
            inverse: vi.fn(),
          }),
        }
      );

      global.DOMPointReadOnly = class {
        x = clientX;
        y = clientY;
        matrixTransform() {
          return {
            x: this.x,
            y: this.y,
          };
        }
      } as any;

      svgElem = document.querySelector("svg") as SVGElement;
      fireEvent.wheel(svgElem, { clientX, clientY, deltaY, ctrlKey: true });
    });

    const viewBoxExpectation = (
      cameraXExpect: number,
      cameraYExpect: number,
      viewWidthExpect: number,
      viewHeightExpect: number
    ) => {
      const [cameraX, cameraY, viewWidthAfterZoom, viewHeightAfterZoom] = (
        svgElem.getAttribute("viewBox") as string
      )
        .split(" ")
        .map((item) => Number(item));
      expect(cameraX).toBeCloseTo(cameraXExpect);
      expect(cameraY).toBeCloseTo(cameraYExpect);
      expect(viewWidthAfterZoom).toBeCloseTo(viewWidthExpect, 5);
      expect(viewHeightAfterZoom).toBeCloseTo(viewHeightExpect, 5);
    };

    it.only("viewport should change properly", async () => {
      viewBoxExpectation(
        cameraXResult,
        cameraYResult,
        viewWidthResult,
        viewHeightResult
      );
    });

    describe.skip("zoom view second time", () => {
      const secondClientX = 200,
        secondClientY = 300,
        secondDeltaY = -10;

      beforeEach(() => {
        fireEvent.wheel(svgElem, {
          secondClientX,
          secondClientY,
          secondDeltaY,
          ctrlKey: true,
        });
      });

      it("viewport should change properly", async () => {
        viewBoxExpectation(
          cameraXResult + secondClientX * 0.1,
          cameraYResult + secondClientY * 0.1,
          viewWidth * 0.7,
          viewHeight * 0.7
        );
      });
    });
  });
});
