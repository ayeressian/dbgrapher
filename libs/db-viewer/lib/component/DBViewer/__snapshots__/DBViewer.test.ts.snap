// Vitest Snapshot v1

exports[`DBViewer should render properly 1`] = `
Object {
  "component": DBViewer {
    "$$": Object {
      "after_update": Array [],
      "before_update": Array [],
      "bound": Object {},
      "callbacks": Object {},
      "context": Map {},
      "ctx": Array [
        Object {
          "annotations": Array [
            Object {
              "height": 600,
              "pos": Object {
                "x": 1200,
                "y": 1200,
              },
              "text": "Description 1",
              "title": "Annotation 1",
              "width": 800,
            },
            Object {
              "height": 200,
              "pos": Object {
                "x": 1400,
                "y": 1100,
              },
              "text": "Description 2",
              "title": "Annotation 2",
              "width": 300,
            },
          ],
          "tables": Array [
            Object {
              "columns": Array [
                Object {
                  "name": "id",
                  "pk": true,
                  "type": "int",
                },
                Object {
                  "name": "cpacity",
                  "type": "int",
                },
                Object {
                  "name": "name",
                  "type": "string",
                },
                Object {
                  "name": "address",
                  "type": "string",
                },
              ],
              "name": "school",
              "pos": Object {
                "x": 722,
                "y": 560,
              },
            },
            Object {
              "columns": Array [
                Object {
                  "name": "id",
                  "pk": true,
                  "type": "int",
                },
                Object {
                  "name": "grade",
                  "type": "int",
                },
                Object {
                  "name": "school_id",
                  "nn": true,
                  "uq": false,
                },
              ],
              "name": "class",
              "pos": Object {
                "x": 620,
                "y": 306,
              },
            },
            Object {
              "columns": Array [
                Object {
                  "name": "id",
                  "pk": true,
                  "type": "int",
                },
                Object {
                  "name": "firstname",
                  "type": "string",
                },
                Object {
                  "name": "lastname",
                  "type": "string",
                },
                Object {
                  "name": "age",
                  "type": "int",
                },
                Object {
                  "fk": Object {
                    "column": "id",
                    "table": "class",
                  },
                  "name": "class_id",
                  "nn": true,
                },
                Object {
                  "fk": Object {
                    "column": "id",
                    "table": "student",
                  },
                  "name": "friend",
                },
              ],
              "name": "student",
              "pos": Object {
                "x": 271,
                "y": 407,
              },
            },
          ],
        },
        1,
        0,
        0,
        <div
          class="svelte-117pbly"
          id="container"
        >
          <svg
            baseProfile="full"
            class="bg-gray-200"
            version="2"
            viewBox="0 0 800 1000"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="svelte-c9fi85"
              d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
            />
             
            <path
              class="highlight svelte-c9fi85"
              d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
            />
            
            
            
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-school"
                height="0"
                width="0"
                x="722"
                y="560"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        school
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        cpacity
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        name
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        address
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-class"
                height="0"
                width="0"
                x="620"
                y="306"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        class
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        grade
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        school_id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        undefined
                      </td>
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-student"
                height="0"
                width="0"
                x="271"
                y="407"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        student
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        firstname
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        lastname
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        age
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        class_id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      />
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        friend
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      />
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
          </svg>
        </div>,
        1000,
        800,
        [Function],
        [Function],
        [Function],
        [Function],
      ],
      "dirty": Array [
        -1,
      ],
      "fragment": Object {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": Array [
        [Function],
      ],
      "on_disconnect": Array [],
      "on_mount": Array [],
      "props": Object {
        "schema": 0,
      },
      "root": <div>
        <div
          class="svelte-117pbly"
          id="container"
        >
          <svg
            baseProfile="full"
            class="bg-gray-200"
            version="2"
            viewBox="0 0 800 1000"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              class="svelte-c9fi85"
              d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
            />
             
            <path
              class="highlight svelte-c9fi85"
              d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
            />
            
            
            
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-school"
                height="0"
                width="0"
                x="722"
                y="560"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        school
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        cpacity
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        name
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        address
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-class"
                height="0"
                width="0"
                x="620"
                y="306"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        class
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        grade
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        school_id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        undefined
                      </td>
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
            
            <g
              class="svelte-5xbau0"
            >
              <foreignobject
                class="svelte-5xbau0"
                data-testid="table-student"
                height="0"
                width="0"
                x="271"
                y="407"
              >
                <table
                  class="svelte-5xbau0"
                >
                  <thead
                    class="svelte-5xbau0"
                  >
                    <tr
                      class="svelte-5xbau0"
                    >
                      <th
                        class="svelte-5xbau0"
                        colspan="3"
                      >
                        student
                      </th>
                    </tr>
                  </thead>
                   
                  <tbody
                    class="svelte-5xbau0"
                  >
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        firstname
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        lastname
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        string
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        age
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        int
                      </td>
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        class_id
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      />
                    </tr>
                    <tr>
                      <td
                        class="svelte-18ks2ry"
                      />
                       
                      <td
                        class="svelte-18ks2ry"
                      >
                        friend
                      </td>
                       
                      <td
                        class="svelte-18ks2ry"
                      />
                    </tr>
                  </tbody>
                </table>
              </foreignobject>
            </g>
          </svg>
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="svelte-117pbly"
        id="container"
      >
        <svg
          baseProfile="full"
          class="bg-gray-200"
          version="2"
          viewBox="0 0 800 1000"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="svelte-c9fi85"
            d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
          />
           
          <path
            class="highlight svelte-c9fi85"
            d="M 290.2116721590354 401.4401751058379 L 620 306 M 270.6058360795177 404.2200875529189 a 1,1 0 1,0 20,0 a 1,1 0 1,0 -20,0"
          />
          
          
          
          
          <g
            class="svelte-5xbau0"
          >
            <foreignobject
              class="svelte-5xbau0"
              data-testid="table-school"
              height="0"
              width="0"
              x="722"
              y="560"
            >
              <table
                class="svelte-5xbau0"
              >
                <thead
                  class="svelte-5xbau0"
                >
                  <tr
                    class="svelte-5xbau0"
                  >
                    <th
                      class="svelte-5xbau0"
                      colspan="3"
                    >
                      school
                    </th>
                  </tr>
                </thead>
                 
                <tbody
                  class="svelte-5xbau0"
                >
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      id
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      cpacity
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      name
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      string
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      address
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      string
                    </td>
                  </tr>
                </tbody>
              </table>
            </foreignobject>
          </g>
          
          <g
            class="svelte-5xbau0"
          >
            <foreignobject
              class="svelte-5xbau0"
              data-testid="table-class"
              height="0"
              width="0"
              x="620"
              y="306"
            >
              <table
                class="svelte-5xbau0"
              >
                <thead
                  class="svelte-5xbau0"
                >
                  <tr
                    class="svelte-5xbau0"
                  >
                    <th
                      class="svelte-5xbau0"
                      colspan="3"
                    >
                      class
                    </th>
                  </tr>
                </thead>
                 
                <tbody
                  class="svelte-5xbau0"
                >
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      id
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      grade
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      school_id
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      undefined
                    </td>
                  </tr>
                </tbody>
              </table>
            </foreignobject>
          </g>
          
          <g
            class="svelte-5xbau0"
          >
            <foreignobject
              class="svelte-5xbau0"
              data-testid="table-student"
              height="0"
              width="0"
              x="271"
              y="407"
            >
              <table
                class="svelte-5xbau0"
              >
                <thead
                  class="svelte-5xbau0"
                >
                  <tr
                    class="svelte-5xbau0"
                  >
                    <th
                      class="svelte-5xbau0"
                      colspan="3"
                    >
                      student
                    </th>
                  </tr>
                </thead>
                 
                <tbody
                  class="svelte-5xbau0"
                >
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      id
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      firstname
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      string
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      lastname
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      string
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      age
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      int
                    </td>
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      class_id
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    />
                  </tr>
                  <tr>
                    <td
                      class="svelte-18ks2ry"
                    />
                     
                    <td
                      class="svelte-18ks2ry"
                    >
                      friend
                    </td>
                     
                    <td
                      class="svelte-18ks2ry"
                    />
                  </tr>
                </tbody>
              </table>
            </foreignobject>
          </g>
        </svg>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
